### ISSUE TITLE
Cache misses in precomputed pow_w in FFT ?

### ISSUE THREAD

**Issue by romainlebreton** [linbox-team/linbox#131]
It seems like calls to the array pow_w are not contiguous (see e.g. [1]) and may cause cache misses. We should use the additional precomputation made in pow_w (see [2]) to solve this problem.

[1] https://github.com/linbox-team/linbox/blob/779556c2a342377ba79fd6b981dfe7fbdafb4b02/linbox/algorithms/polynomial-matrix/polynomial-fft-transform.inl#L67

[2] https://github.com/linbox-team/linbox/blob/779556c2a342377ba79fd6b981dfe7fbdafb4b02/linbox/algorithms/polynomial-matrix/polynomial-fft-transform.h#L91

**Comment by BastienVialla** [comment]
Indeed, we can improve the cache locality, and thus the performances, of the DIF and DIT by reordering the pow_w in bit reverse order (see [1], very dumb and inefficient implementation).

We can reorder the for loops of the DIF (see [2]) and the DIT (see [3]) such that the access to the pow_w vector is an invariant of the inner loop, hence improving cache locality.

Here some benches comparing the performances of the DIF and DIT for modular<uint64_t, uint128_t> and modular<uint32_t, uint64_t>, done on an Intel Xeon E5-2620 v4 (see [4] for the bench file).

We can see that DIF_sort and DIT_sort always performs better than DIF_no_simd and DIT_no_simd.
Futhermore, their performances remains stable while the performances of the other versions decrease as the size of the convolution increase.

The fact that the performances of the no_simd versions decrease earlier for uint64 than for uint32 seems to indicate that the problem come from cache misses.

It might be worth to look if those modifications can be add to the SIMD versions.

![int64](https://user-images.githubusercontent.com/4585751/45211716-49080880-b293-11e8-8487-da64ded2a032.png)
![int32](https://user-images.githubusercontent.com/4585751/45211710-44dbeb00-b293-11e8-9fff-0efe2a274ba3.png)
![int32_sse](https://user-images.githubusercontent.com/4585751/45211712-46a5ae80-b293-11e8-8f89-7e3d785fdd93.png)


[1] https://github.com/linbox-team/linbox/blob/better_fft/linbox/algorithms/polynomial-matrix/polynomial-fft-init.h#L272

[2] https://github.com/linbox-team/linbox/blob/better_fft/linbox/algorithms/polynomial-matrix/polynomial-fft-algorithms.h#L92

[3] https://github.com/linbox-team/linbox/blob/better_fft/linbox/algorithms/polynomial-matrix/polynomial-fft-algorithms.h#L106

[4] https://github.com/linbox-team/linbox/blob/better_fft/examples/bench-fft-2.C

**Comment by romainlebreton** [comment]
I have the feeling that the one place where pow_w and pow_wp were not read contiguously was in FFT NoSimd. I have fixed this in commit [d7aeabc](https://github.com/linbox-team/linbox/commit/d7aeabc1f432acd6853a2e16e9d37ea3b1a2207b) by using the precomputations that are done in pow_w = [1, w, w^2, .... , 1, w^2, w^4, ..., 1, w^4, w^8, ...].

@BastienVialla I don't know why you would talk about pow_w in the bitreverse order. In our butterflies, the powers of w are in the natural order. For example, look at the code in polynomial-fft-algorithms.h, you can see that `DIF_mod2p` in the `class FFT_algorithms<Field, simd, 4> : public FFT_butterflies<Field, simd, 4>` reads contiguously the powers of w.

**Comment by romainlebreton** [comment]
@BastienVialla Ok, I think I understand now why you would talk about pow_w in the bitreverse order. There is indeed a FFT variant that uses pow_w in the bitreverse order. However, we did not choose this variant and our variant does not suffer from cache misses after all.
